### PR TITLE
Find cryptographic materials during a Goat Rodeo run

### PR DESCRIPTION
**Pull request by dpp**
Adds the Certificates strategy (X.509, SSH, PGP, private keys, keystores, CRLs) with full parsing, pURL emission, and metadata extraction. Includes 746 test corpus fixtures checked into git, 35 new test/support Scala files, and removes all Python/shell/Java corpus generation scripts — the test suite depends only on checked-in fixtures and the JVM.
Key changes
- Certificates strategy (Certificates.scala, CertificatesState.scala, CertificatesOidMaps.scala): X.509 cert/CRL/keystore/bundle parsing, SSH pubkey/cert parsing, PGP key parsing, private key classification, pURL construction, and metadata emission
- pURL construction (partially migrated to PackageURLBuilder): SSH, private key, and SPKI pURLs use the builder; purlForPgpKey, purlsForCert, and purlForCrl still use string concatenation and crash on qualifier values containing </> (see PurlConstructionTests)
- Test corpus: 746 fixtures (X.509 roots/synthetic/PQC, SSH keys/certs, PGP keys, private keys, keystores, CRLs, edge cases) with sidecars checked into git — no Python or external tooling required at test time
- Removed 4,700 lines of corpus generation scripts (bootstrap_*.py, generate.sh, Generate*.java, cert_sidecar.py, openssh_v1_fingerprint.py)
- Helpers: CryptoDetector for MIME classification, SshWireFormat for OpenSSH wire parsing, computeSHA256/sha256Hex in Helpers
- 1296 tests passing including property-based, corpus integrity, white-box ground-truth, and adversarial pURL construction tests

### CODE REVIEW

File: info/architecture.md [hunk 1]
```diff
@@ -295,6 +306,19 @@ val strategies = Vector(
 )
 ```
 
+**Reference example:** the `Certificates` strategy
```
Comment (aredridel) [inline info/architecture.md:309]: Future inclusion: netscape key stores. And eventually, windows whatever. That'll be a mess.

File: info/goat_rodeo_operation.md [hunk 2]
```diff
@@ -82,6 +86,19 @@ The `MavenToProcess` looks for groupings of `pom`, `jar`, `sources`, and `javado
...
 associates a source file with a` `.class` file based on the class file's debug information, etc.
 
+The `Certificates` strategy is notable for two design choices that
+differ from Maven/Docker:
+
+- **Keystores produce a single flat Item** with all contained pURLs
+  and metadata aggregated, not one Item per entry. Maven groups
+  `(pom, jar, sources, javadocs)` into one Item-with-children;
+  Certificates aggregates all entries in a JKS / PKCS#12 / etc.
+  into one Item with no children.
```
Comment (aredridel) [inline info/goat_rodeo_operation.md:96]: Interesting choice: why?

I'd think that "find this compromised CA key" is a core usecase for us, by signature of the cert or DN

File: info/goat_rodeo_operation.md [hunk 3]
```diff
@@ -82,6 +86,19 @@ The `MavenToProcess` looks for groupings of `pom`, `jar`, `sources`, and `javado
...
+- **Encrypted material is opaque.** Encrypted private keys and
+  encrypted keystores produce envelope-only metadata (KDF, cipher,
+  salt) and no pURL. The strategy never attempts decryption, never
+  guesses passwords, never logs encrypted content. 
+
```
Comment (aredridel) [inline info/goat_rodeo_operation.md:101]: This makes sense. Can we get signatures from common key types?

File: info/mime_types.md [hunk 4]
```diff
@@ -21,6 +21,65 @@ At present there are two post-processing operations that refine MIME types:
...
+augmenter (registered after Saffron in
+`ArtifactWrapper.augmenters`) inspects the first 4 KB of each
+artifact and ADDS specific MIME types based on content signatures.
+
+**Pure-addition design:** unlike `SaffronDetector`, which strips
+`text/*` MIMEs when it finds a binary signature, `CryptoDetector` is
+purely additive — it never removes a Tika-detected MIME, only adds
+its own. Rationale: PEM-encoded artifacts legitimately are BOTH
+`text/plain` (Tika) AND `application/x-pem-file` (CryptoDetector);
+downstream filtering chooses what to act on.
```
Comment (aredridel) [inline info/mime_types.md:38]: This is an interesting principle and something we might want to do more of. Strategies become lenses, and we can use multiple lenses to look at things.

File: info/mime_types.md [hunk 5]
```diff
@@ -21,6 +21,65 @@ At present there are two post-processing operations that refine MIME types:
...
+| `-----BEGIN PGP MESSAGE-----` | `application/pgp-message` |
+| Multi-block PEM with ≥ 2 `BEGIN CERTIFICATE` blocks | `application/x-pem-file` + `application/x-pem-bundle` |
+| SSH wire-format public-key line (`ssh-rsa AAAAB3…`, `ssh-ed25519 …`, `ecdsa-sha2-nistpXXX …`, `ssh-dss …`, `sk-ssh-ed25519@openssh.com …`, `sk-ecdsa-sha2-nistp256@openssh.com …`) | `application/x-openssh-public-key` |
+| OpenSSH cert wire-format line (`ssh-{rsa,dss,ed25519,ed448}-cert-v01@openssh.com`, `ecdsa-sha2-nistp{256,384,521}-cert-v01@openssh.com`) | `application/x-openssh-certificate` |
+| DER X.509 cert (probe via BC X.509 parser) | `application/pkix-cert` |
+| DER X.509 CRL (probe via BC CRL parser) | `application/pkix-crl` |
+| DER PKCS#7 SignedData (OID 1.2.840.113549.1.7.2 near start) | `application/pkcs7-mime` |
+| `0xfe 0xed 0xfe 0xed` magic | `application/x-java-keystore` (JKS) |
+| `0xce 0xce 0xce 0xce` magic | `application/x-java-jce-keystore` (JCEKS) |
+| `0x30 0x82` DER-prefix + `.p12` / `.pfx` extension | `application/pkcs12` |
```
Comment (aredridel) [inline info/mime_types.md:71]: Can we detect raw DER encoded keys?

File: info/mime_types.md [hunk 6]
```diff
@@ -21,6 +21,65 @@ At present there are two post-processing operations that refine MIME types:
...
+| `-----BEGIN X509 CRL-----` | `application/x-pem-file` + `application/pkix-crl` |
+| `-----BEGIN PKCS7-----` / `-----BEGIN CMS-----` | `application/pkcs7-mime` |
+| `-----BEGIN PUBLIC KEY-----` | `application/x-pem-file` + `application/x-pem-public-key` |
+| `-----BEGIN RSA/EC/DSA PRIVATE KEY-----` / `-----BEGIN PRIVATE KEY-----` | `application/x-pem-file` + `application/x-pem-private-key` |
+| `-----BEGIN ENCRYPTED PRIVATE KEY-----` | `application/x-pem-file` + `application/x-pem-encrypted-private-key` |
+| `-----BEGIN OPENSSH PRIVATE KEY-----` | `application/x-openssh-private-key` |
+| `-----BEGIN PGP PUBLIC KEY BLOCK-----` / `PRIVATE KEY BLOCK` | `application/pgp-keys` |
+| `-----BEGIN PGP SIGNATURE-----` | `application/pgp-signature` |
+| `-----BEGIN PGP MESSAGE-----` | `application/pgp-message` |
+| Multi-block PEM with ≥ 2 `BEGIN CERTIFICATE` blocks | `application/x-pem-file` + `application/x-pem-bundle` |
```
Comment (aredridel) [inline info/mime_types.md:63]: NEAT

File: info/mime_types.md [hunk 7]
```diff
@@ -21,6 +21,65 @@ At present there are two post-processing operations that refine MIME types:
...
+| SSH wire-format public-key line (`ssh-rsa AAAAB3…`, `ssh-ed25519 …`, `ecdsa-sha2-nistpXXX …`, `ssh-dss …`, `sk-ssh-ed25519@openssh.com …`, `sk-ecdsa-sha2-nistp256@openssh.com …`) | `application/x-openssh-public-key` |
+| OpenSSH cert wire-format line (`ssh-{rsa,dss,ed25519,ed448}-cert-v01@openssh.com`, `ecdsa-sha2-nistp{256,384,521}-cert-v01@openssh.com`) | `application/x-openssh-certificate` |
+| DER X.509 cert (probe via BC X.509 parser) | `application/pkix-cert` |
+| DER X.509 CRL (probe via BC CRL parser) | `application/pkix-crl` |
+| DER PKCS#7 SignedData (OID 1.2.840.113549.1.7.2 near start) | `application/pkcs7-mime` |
+| `0xfe 0xed 0xfe 0xed` magic | `application/x-java-keystore` (JKS) |
+| `0xce 0xce 0xce 0xce` magic | `application/x-java-jce-keystore` (JCEKS) |
+| `0x30 0x82` DER-prefix + `.p12` / `.pfx` extension | `application/pkcs12` |
+| Binary PGP packet — first byte ∈ {0xC6 (new-format tag-6), 0x98, 0x99, 0x9A, 0x9B (old-format tag-6 length variants)} | `application/pgp-keys` |
```
Comment (aredridel) [inline info/mime_types.md:72]: Single byte? eek

File: src/main/scala/io/spicelabs/goatrodeo/omnibor/strategies/Certificates.scala [hunk 8]
```diff
@@ -0,0 +1,1894 @@
...
+
+  /** A successfully-parsed artifact ready for emission. The variant tells
+    * `getPurls` / `getMetadata` how to dispatch.
+    */
+  sealed trait ClaimedContent
+
+  /** Single X.509 certificate. */
+  final case class SingleCert(cert: X509Certificate) extends ClaimedContent
+
+  /** Java/BC keystore.
```
Comment (aredridel) [inline src/main/scala/io/spicelabs/goatrodeo/omnibor/strategies/Certificates.scala:87]: 📓 BC = BouncyCastle

File: src/main/scala/io/spicelabs/goatrodeo/omnibor/strategies/Certificates.scala [hunk 9]
```diff
@@ -0,0 +1,1894 @@
...
+  private[strategies] def parseBundle(
+      artifact: ArtifactWrapper
+  ): Option[Bundle] = {
+    val collected: Option[Vector[X509Certificate]] = artifact.withStream { f =>
+      Try {
+        Using.resource(
+          new PEMParser(
+            new InputStreamReader(
+              f,
+              "ISO-8859-1"
```
Comment (aredridel) [inline src/main/scala/io/spicelabs/goatrodeo/omnibor/strategies/Certificates.scala:344]: Why ISO-8859-1? Does this accept the bytes that 8859-1 rejects?

File: src/main/scala/io/spicelabs/goatrodeo/omnibor/strategies/Certificates.scala [hunk 10]
```diff
@@ -0,0 +1,1894 @@
...
+    "ssh-rsa" -> SshAlgMap("rsa", None, false),
+    "ssh-dss" -> SshAlgMap("dsa", Some(("size", "1024")), false),
+    "ssh-ed25519" -> SshAlgMap("ed25519", None, false),
+    "ssh-ed448" -> SshAlgMap("ed448", None, false),
+    "ecdsa-sha2-nistp256" -> SshAlgMap("ec", Some(("curve", "p-256")), false),
+    "ecdsa-sha2-nistp384" -> SshAlgMap("ec", Some(("curve", "p-384")), false),
+    "ecdsa-sha2-nistp521" -> SshAlgMap("ec", Some(("curve", "p-521")), false),
+    "sk-ssh-ed25519@openssh.com" -> SshAlgMap("ed25519", None, true),
+    "sk-ecdsa-sha2-nistp256@openssh.com" -> SshAlgMap(
+      "ec",
```
Comment (aredridel) [inline src/main/scala/io/spicelabs/goatrodeo/omnibor/strategies/Certificates.scala:426]: from my SSH connection: 

```host key algorithms: ssh-ed25519-cert-v01@openssh.com,ecdsa-sha2-nistp256-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,sk-ssh-ed25519-cert-v01@openssh.com,sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ssh-ed25519@openssh.com,sk-ecdsa-sha2-nistp256@openssh.com,rsa-sha2-512,rsa-sha2-256```

"Wire algorithm" here is which set of things?

File: src/main/scala/io/spicelabs/goatrodeo/omnibor/strategies/CertificatesOidMaps.scala [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+/** Canonical-mapping tables for the Certificates strategy. Pure data — no
+  * logic, no side effects. Maps are `private[strategies]` for use by
+  * `Certificates` and `CertificatesState`.
+  */
+private[strategies] object CertificatesOidMaps {
+
+  // ---------- X.509 algorithm + signature OIDs -----------
+
+  /** Public-key OIDs → (canonical alg, optional params token). */
```
Comment (aredridel) [inline src/main/scala/io/spicelabs/goatrodeo/omnibor/strategies/CertificatesOidMaps.scala:25]: Origin of this data would be a good comment to have, so we can track updates more easily

File: src/main/scala/io/spicelabs/goatrodeo/util/CryptoDetector.scala [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+    val limit = math.min(untilIndex, haystack.length) - needle.length + 1
+    var i = fromIndex
+    while (i < limit) {
+      var j = 0
+      while (j < needle.length && haystack(i + j) == needle(j)) j += 1
+      if (j == needle.length) return true
+      i += 1
+    }
+    false
+  }
```
Comment (aredridel) [inline src/main/scala/io/spicelabs/goatrodeo/util/CryptoDetector.scala:379]: Not a library function?

File: src/test/scala/strategies/CertificatesSidecarTests.scala [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+limitations under the License. */
+
+package strategies
+
+import munit.FunSuite
+
+import java.io.File
+import java.nio.file.Files
+
+/** Unit tests for [[CertificatesSidecar.parse]].
```
Comment (aredridel) [inline src/test/scala/strategies/CertificatesSidecarTests.scala:22]: Is this really tests for the test suite?

**Review by aredridel — APPROVED** [review]

File: info/architecture.md (hunk 1 above)
Comment (dpp) [reply info/architecture.md:309]: Sound like good expantion

File: info/goat_rodeo_operation.md (hunk 2 above)
Comment (dpp) [reply info/goat_rodeo_operation.md:96]: It avoids treating the key container like a container which would have radically expanded the code requirements (both Strategy and FileWalker).

We do get a separate pURL for each key found in the key container, so I think the use case still works... we can search for the key based on pURL.

File: info/goat_rodeo_operation.md (hunk 3 above)
Comment (dpp) [reply info/goat_rodeo_operation.md:101]: I am trying to avoid even the remotest possibility of leaking any key material... it's probably possible, but the cost of a mistake here is high.

File: info/mime_types.md (hunk 4 above)
Comment (dpp) [reply info/mime_types.md:38]: Mayyyybeeee.... let's see where there are strategy overlaps... so far they are Java, linux/open bsd packages, docker images, certs, and one more... haven't seen an overlap.

File: info/mime_types.md (hunk 5 above)
Comment (dpp) [reply info/mime_types.md:71]: I believe so

File: info/goat_rodeo_operation.md (hunk 3 above)
Comment (aredridel) [reply info/goat_rodeo_operation.md:101]: Makes some sense

File: src/main/scala/io/spicelabs/goatrodeo/omnibor/strategies/Certificates.scala (hunk 9 above)
Comment (dpp) [reply src/main/scala/io/spicelabs/goatrodeo/omnibor/strategies/Certificates.scala:344]: I think a lot of certs, etc. were ISO-8859-1 encoded... could be wrong.

If there are illegal characters, the function will throw and the item will not be decoded.

If we need to support multiple encodings, it's possible to try one after another.

File: info/goat_rodeo_operation.md (hunk 2 above)
Comment (aredridel) [reply info/goat_rodeo_operation.md:96]: Hm. I couldn't easily find an example purl in the test data — what does the purl for a specific cert look like? Rather than just "there is a cert of this kind"?

File: src/main/scala/io/spicelabs/goatrodeo/util/CryptoDetector.scala (hunk 12 above)
Comment (dpp) [reply src/main/scala/io/spicelabs/goatrodeo/util/CryptoDetector.scala:379]: Prolly should be. But which library and is the dependency worth it? Thoughts?

Comment (aredridel) [reply src/main/scala/io/spicelabs/goatrodeo/util/CryptoDetector.scala:379]: If it needs a dep, nah, if it's builtin, yeah.

File: src/test/scala/strategies/CertificatesSidecarTests.scala (hunk 13 above)
Comment (dpp) [reply src/test/scala/strategies/CertificatesSidecarTests.scala:22]: Yep... belt and suspenders. There were actually tests that ensured particular content in the documentation... I yanked those...
